src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1029 @@
+}
+
+void tandberg_tdv2100_disp_logic_device::vblank(int state)
+{
+	if(state && !m_vblank_state)
+	{
+		// TODO: VSync interrupt for CPU interface
+		m_frame_counter++;
+		m_screen->update_now();
+	}
+	m_vblank_state = state;
+}
+
+int tandberg_tdv2100_disp_logic_device::get_attribute(int at_row, int at_col)
+{
+	int attribute = 0;
+	for(int char_row = 0; char_row <= at_row; char_row++)
+	{
+		for(int char_col = 0; char_col < 80; char_col++)
+		{
+			int const vram_data = m_vram[char_col + char_row*80];
+			if(vram_data&0x80)
+			{
+				attribute = (vram_data&0x70)>>4;
+			}
+			if(char_row == at_row && char_col == at_col)
+			{
+				break;
+			}
+		}
+	}
+	return attribute;
+}
+
+uint32_t tandberg_tdv2100_disp_logic_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	rgb_t const *const palette = m_palette->palette()->entry_list_raw();
+
+	bool extend_dot7 = ((m_dsw_u61->read()&0x003) == 1);
+	bool blinking_cursor = ((m_dsw_u61->read()&0x030) == 0x010);
+	bool blank_attr_chars = !(m_dsw_u61->read()&0x040);
+	bool blank_ctrl_chars = !(m_dsw_u61->read()&0x100);
+	bool block_cursor = m_dsw_u61->read()&0x200;
+	bool full_inverse_video = !m_sw_invert_video->read();
+
+	int const char_min_col = cliprect.min_x/9;
+	int const char_max_col = cliprect.max_x/9;
+	int const char_min_row = cliprect.min_y/14;
+	int const char_max_row = cliprect.max_y/14;
+	bool const blink_strobe = m_frame_counter&0x08;
+	bool const cursor_blink_strobe = blink_strobe && !m_speed_check;
+
+	for(int char_row_nr = char_min_row; char_row_nr <= char_max_row; char_row_nr++)
+	{
+		// This is at the start of a char row
+		int const chr_y_pos = char_row_nr*14;
+		int attribute = get_attribute(char_row_nr, char_min_col);
+
+		for (int char_col_nr = char_min_col; char_col_nr <= char_max_col; char_col_nr++)
+		{
+			// This is at the start of a single character
+			int const chr_x_pos = char_col_nr*9;
+			int const in_line_char_nr = char_col_nr - 4;
+			int const vram_address = get_ram_addr(char_row_nr, in_line_char_nr);
+			bool const is_cursor = (m_cursor_col == in_line_char_nr && m_cursor_row == char_row_nr);
+
+			for(int line = 0; line < 14; line++)
+			{
+				// This is at the start of a 9-dot raster-line segment in a single character
+				uint8_t data = 0x00;
+				int dot_nr8 = 0;
+				int extra_intensity = 1;
+				if(in_line_char_nr >= 0 && in_line_char_nr < 80)
+				{
+					bool const draw_cursor = (is_cursor && (line == 12 || block_cursor) && (!blinking_cursor || cursor_blink_strobe));
+					int const vram_data = m_vram[vram_address];
+
+					// Pattern generator & character blanking
+					if(m_video_enable && (vram_data&0x60 || !blank_ctrl_chars) && !(vram_data&0x80 && blank_attr_chars))
+					{
+						data = m_font[((vram_data&0x7f)<<4) + line];
+						if(extend_dot7)
+						{
+							dot_nr8 = data&0x01;
+						}
+					}
+
+					// Pattern modifiers
+					bool underline_mode = m_dsw_u61->read()&0x008;
+					if(underline_mode)
```

Comment:
You’ve got a read of an I/O port on every line of every character here.

Reply:
This slipped under the radar, moved out of the loops now.